PSPs/drafts/psp-34.md

```diff
@@ -0,0 +1,914 @@
+        "type": "Option<AccountId>"
+      }
+    },
+    {
+      "docs": [],
+      "indexed": true,
+      "label": "to",
+      "type": {
+        "displayName": [
+          "Option"
+        ],
+        "type": "Option<AccountId>"
+      }
+    },
+    {
+      "docs": [],
+      "indexed": true,
+      "label": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    }
+  ],
+  "docs": [
+    "Event emitted when a token transfer occurs."
+  ],
+  "label": "Transfer"
+}
+```
+
+### Approval
+```json
+{
+  "args": [
+    {
+      "docs": [],
+      "indexed": true,
+      "label": "from",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "docs": [],
+      "indexed": true,
+      "label": "to",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "docs": [],
+      "indexed": true,
+      "label": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    }
+  ],
+  "docs": [
+    "Event emitted when a token approve occurs."
+  ],
+  "label": "Approval"
+}
+```
+
+### ApprovalForAll
+```json
+{
+  "args": [
+    {
+      "docs": [],
+      "indexed": true,
+      "label": "owner",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "docs": [],
+      "indexed": true,
+      "label": "operator",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "docs": [],
+      "indexed": false,
+      "label": "approved",
+      "type": {
+        "displayName": [
+          "bool"
+        ],
+        "type": "bool"
+      }
+    }
+  ],
+  "docs": [
+    "Event emitted when an operator is enabled or disabled for an owner.",
+    "The operator can manage all NFTs of the owner."
+  ],
+  "label": "ApprovalForAll"
+}
+```
+
+### AttributeSet
+```json
+{
+  "args": [
+    {
+      "docs": [],
+      "indexed": false,
+      "label": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    },
+    {
+      "docs": [],
+      "indexed": false,
+      "label": "key",
+      "type": {
+        "displayName": [
+          "[u8]"
+        ],
+        "type": "[u8]"
+      }
+    },
+    {
+      "docs": [],
+      "indexed": false,
+      "label": "data",
+      "type": {
+        "displayName": [
+          "[u8]"
+        ],
+        "type": "[u8]"
+      }
+    }
+  ],
+  "docs": [
+    "Event emitted when an attribute is set for a token.",
+  ],
+  "label": "AttributeSet"
+}
+```
+
+### Types
+```rust
+// Id is an Enum and its variant are types
+enum Id {
+    U8(u8),
+    U16(u16),
+    U32(u32),
+    U64(u64),
+    U128(u128),
+    Bytes(Vec<u8>),
+}
```

Comment:
does it make sense to have `Id` as enumerable?
can't we just make it fixed-size bytes array that will be allocated anyways?

Reply:
Then the question which fixed-size we need to use?=)

Reply:
I think 32 bytes as for AccountId

Reply:
But for many contracts, it will be enough to have u64 as an identifier of the NFT. In some cases, it can be less, in some cases, it should be more. It depends on the implementation. So it is why we decided to create it like enumerable because WASM compatible languages can implement that

Reply:
if we have it as enum, does that mean we will always allocate memory for the biggest variant (in our case 3 words for Vec)?